### PR TITLE
Customize scroll behavior

### LINKED ISSUE #2884: Feature Request: Ability to customize scroll behaviour in WebChat V4
**Is your feature request related to a problem? Please describe.**

User experience is one of the most important KPI’s for our entire website and thus for our product. This includes the ease at which users can read the information the chatbot provides them with. To this end, we feed our users small chunks of text. Most responses consist of multiple text bubbles. By default, the WebChat scrolling behavior is in such a way that it automatically scrolls all the way down to the most recent text bubble. Since we compose responses consisting of multiple separate text bubbles, the user interface ‘bounces’ down with every text bubble, ruining our user’s experience. This is a problem, because it is paramount that our users can take their time reading the information, and empower them to scroll down themselves when they’re ready. We require scrolling behaviour that enables this ease of reading while visually making clear that new messages are available. We are quite surprised that WebChat V4 does not yet support a feature to customize the scrolling behavior.

**Describe the suggestion or request in detail**

A feature or customization option (Hook) that lets us control the client’s scrolling behavior. We require that both the last user input, the first bot response to that user message, and 2,5 lines of text of the 2nd chatbot text bubble are in view of the client. A setting or Hook that would allow us to customise scrolling behavior to our own requirements would be very much appreciated.

**Describe alternatives you have considered**

- continue using WebChat V3 while it still functions; we have built this feature ourselves for WebChat V3. This is not acceptable because we need to keep up to date with technologies.
- taking our chatbot offline when WebChat V3 stops working as long as WebChat V4 doesn’t support this feature. This is not acceptable because we have a duty to provide services with our product.
- using another chat vendor that does support this customization feature.

**Additional context**

We are currently using WebChat V3, but would very much like to upgrade to V4 when it includes this customization feature.

To clearify our request we uploaded a video, you can find the video [here](https://youtu.be/-V9cdeif90c)

[Enhancement]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #2884.

### Added

-  Added new style options to customize auto-scroll, by [@compulim](https://github.com/compulim) in PR [#3653](https://github.com/microsoft/BotFramework-WebChat/pull/3653)
   -  Set `autoScrollSnapOnActivity` to `true` to pause auto-scroll after more than one activity is shown, or a number to pause after X number of activities
   -  Set `autoScrollSnapOnPage` to `true` to pause auto-scroll when a page is filled, or a number between `0` and `1` to pause after % of page is filled
   -  Set `autoScrollSnapOnActivityOffset` and `autoScrollSnapOnPageOffset` to a number (in pixels) to overscroll/underscroll after the pause
-  Supports multiple transcripts in a single composition, by [@compulim](https://github.com/compulim) in PR [#3653](https://github.com/microsoft/BotFramework-WebChat/pull/3653)

## Description

Added style options to customize scroll behavior by pausing at specified "snap point".

Currently, support 2 snap points:

- Pause after a percentage of page is filled
- Pause after X activities arrived

In additional to snap points, 2 more options to specify number of pixels to overscroll or underscroll. If both options are specified, it will use the value that would pause the auto-scroll *sooner*.

The following video demonstrates the pause after 2 activities and 100 pixels. The cyan color indicates acknowledged activities and orange color indicates recently arrived activities. Equivalent to setting:

```js
styleOptions = {
  autoScrollPauseOnActivity: 2,
  autoScrollPauseOnActivityOffset: 100
};
```

https://user-images.githubusercontent.com/1622400/103579647-83970580-4e8d-11eb-942f-15110c511191.mp4

### Default behavior

By default, we did not enable this feature. We should consult our UX team before turning it on by default.

### Activities acknowledgement

To pause, only unacknowledged activities are taken into account. For example, the page will pause after it filled with activities the user didn't positively indicate they have been read.

Acknowledgement means: activity is positively acknowledged by the user. If the user does not interact on Web Chat, no activities should be acknowledged.

Currently, we use 2 signals for indicating acknowledgement:

- When the user scroll to the bottom of the page
- When the user sent a message

The signals above positively identify the user read the content.

Since the DLJS and other chat adapters do not keep this acknowledgement information, we will assume, when Web Chat started, all activities are acknowledged (e.g. restoring from conversation history). And it will stop auto-acknowledging when the user send their first message.

Due to the limitation of chat adapter, we recommend content author not to send initial activities (a.k.a. welcome messages) longer than a page.

### Allowing multiple instances of `<BasicTranscript>`

Last October, when we work on the feature to refactor some React components into API layer, the way we refactor `<BasicTranscript>` does not allow multiple instances of it inside the same composition (i.e. `<Composer>`).

In this PR, we also addressed it by allowing multiple instances of `<BasicTranscript>`, such as:

```js
<Composer directLine={...}>
  <BasicTranscript />
  <BasicTranscript />
</Composer>
```

### Resolving #2884 

In #2884, the developer prefers to pause "after 1 activity and 2.5 lines of text". Since line of text is not trivial to measure in HTML, we fallback to pixels instead.

To resolve #2884, the developer should set the following style options:

```js
styleOptions = {
  autoScrollPauseOnActivity: true,
  autoScrollPauseOnActivityOffset: 100
};
```

## Specific Changes

- Modified `<Composer>` and corresponding hooks to support multiple instances of `<BasicTranscript>`
  - This is done by multiplexing several hooks
  - For example, when `useScrollTo` is being called, we will multiplex this call into all instances of `<BasicTranscript>`, instead of one
- Added 4 style options: `autoScrollSnapOnActivity`, `autoScrollSnapOnActivityOffset`, `autoScrollSnapOnPage`, `autoScrollSnapOnPageOffset`
- Bumped `react-scroll-to-bottom@4.1.0` to support pausing at a specific `scrollTop` position (https://github.com/compulim/react-scroll-to-bottom/pull/73)

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] Browser and platform compatibilities reviewed
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] Security reviewed (no data URIs, check for nonce leak)
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: __tests__/html/autoScroll.acknowledgement.html
Comment (compulim) [reply __tests__/html/autoScroll.acknowledgement.html:145]: The test has a fixed resolution when it is running under Jest. If you are opening the HTML, your screen resolution may not match its expectation.

`x.scrollTop === y.offsetTop` means, the element `y` is showed at the top of the scrollable `x`.

File: __tests__/html/autoScroll.snap.default.html
Comment (compulim) [reply __tests__/html/autoScroll.snap.default.html:91]: Are you running the tests under Docker (i.e. via Jest)?

File: packages/api/src/defaultStyleOptions.js
Comment (corinagum) [inline packages/api/src/defaultStyleOptions.js:206]:
```suggestion
  autoScrollSnapOnActivity: false, // true will pause scroll after 1 activity is received. Specifying a number will pause after X number of activities
```

File: packages/api/src/defaultStyleOptions.js
Comment (corinagum) [inline packages/api/src/defaultStyleOptions.js:208]:
```suggestion
  autoScrollSnapOnPage: false, // true will pause scroll after activities filled the page. Specifying a number (0 to 1) will pause after % of page is filled
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:13]:
```suggestion
//    - We don't need a condition here. When Web Chat sends the message, it will scroll to bottom, and it will trigger condition 2 below.
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:15]:
```suggestion
//    - If the transcript is already at the bottom, the user needs to scroll up and then go back down
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:16]:
```suggestion
//    - What happens if we are relaxing "scrolled from a non-bottom scroll position":
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:14]: I'm a little confused by this section / the intent. 

This line seems to address to different scenarios. 
Bottom of the transcript = acknowledged. 

"scrolled from a non-bottom scroll position" - I'm assuming this refers to manual scrolling by the user, is that correct?

"Activity is acknowledged when user is already at the bottom of the transcript or user manually scrolls down to the bottom of the transcript" -- does this better summarize this line?

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:15]: Is this the new behavior?

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:22]:
```suggestion
// Note: When Web Chat is loaded, there are no activities acknowledged. We need to assume all arriving activities are acknowledged until end-user sends their first activity.
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:26]: Add to list of TODOs

File: packages/component/src/Utils/findLastIndex.js
Comment (corinagum) [inline packages/component/src/Utils/findLastIndex.js:4]: why `~index` bitwise NOT?

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:16]: So are you saying that with this PR, the new behavior is 'relaxed scrolled from a non-bottom scroll position'?
Don't use future tense in that sense - if this is the commit that's making changes, it needs to be phrased in present tense.

This seems more like commentary that should go in change description, not necessarily as code comments (e.g. in PR summary/release summary). What do you think?

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:20]: add "therefore... + explanation"

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:24]:
```suggestion
//       It would be better if the chat adapter should let Web Chat know if the activity is loaded from history or not.
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:52]:
```suggestion
    // As described above, if no activities were acknowledged through egress activity, we will assume everything is acknowledged.
```

This is for PoliCheck

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:60]: remove `stickyChangedToSticky`

File: __tests__/html/autoScroll.snap.activity.html
Comment (corinagum) [inline __tests__/html/autoScroll.snap.activity.html:39]: delete

File: __tests__/html/autoScroll.acknowledgement.html
Comment (corinagum) [inline __tests__/html/autoScroll.acknowledgement.html:51]: delete

File: __tests__/html/autoScroll.snap.activityAndPage.html
Comment (corinagum) [inline __tests__/html/autoScroll.snap.activityAndPage.html:39]: delete

File: __tests__/html/autoScroll.snap.default.html
Comment (corinagum) [inline __tests__/html/autoScroll.snap.default.html:39]: delete

File: __tests__/html/autoScroll.snap.page.html
Comment (corinagum) [inline __tests__/html/autoScroll.snap.page.html:39]: delete

File: packages/testharness/src/utils/createDirectLineWithTranscript.js
Comment (corinagum) [inline packages/testharness/src/utils/createDirectLineWithTranscript.js:28]: Just posting the activity multiple times, right?

File: packages/component/src/hooks/useScrollToEnd.js
Comment (corinagum) [inline packages/component/src/hooks/useScrollToEnd.js:8]: is smooth behavior because we now have 'jump scroll to' as well?

File: packages/component/src/hooks/useScrollTo.js
Comment (corinagum) [inline packages/component/src/hooks/useScrollTo.js:8]: so does this count as jump scroll to?

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [inline packages/component/src/hooks/internal/useAcknowledgedActivity.js:36]: Why can't we use `useMemo` here?

Comment (corinagum) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:36]: oh is it because it becomes nested?

File: packages/testharness/src/utils/createDirectLineWithTranscript.js
Comment (corinagum) [inline packages/testharness/src/utils/createDirectLineWithTranscript.js:66]: don't need echo - always true

**Review by corinagum — APPROVED** [review]
LGTM - I approved (with comments), but since @amal-khalaf is reviewing as well, I recommend waiting for her approval too.

File: __tests__/html/autoScroll.snap.default.html
Comment (amal-khalaf) [reply __tests__/html/autoScroll.snap.default.html:90]: yes

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (compulim) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:14]: "scrolled from a non-bottom scroll position", yes, this is manual scrolling by the user. I have updated it to "The user scroll to the bottom of the transcript, from a non-bottom scroll position".

Your statement is not exactly correct because of this part "activity is acknowledged when user is at the bottom of the transcript".

Consider this, the developer set "pause after 3 activities". And Web Chat received 1 single line activity, and another after 5 seconds (proactive), then nothing else. Sum it up: we received 2 activities, they are short, and the transcript is still sticky.

These 2 activities should not be acknowledged even though the user is at the bottom of the transcript.

If we acknowledged them, assume 2 more single-liner come in (total 4). Because the first 2 were already acknowledged, scroll will not pause when last 2 come in. Then we violated "pause after 3 activities".

I rewrote this part 3 times because I keep getting burnt by these "when to pause" algorithms. 🤦🏻‍♂️

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (compulim) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:20]: Added:

```js
//      5. Therefore, even the developer set "pause after 3 activities", if activities are coming in at a slow pace (not batched in a single render)
//         Web Chat will keep scrolling and not snapped/paused
```

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (compulim) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:26]: Added new issue #3670.

File: packages/component/src/hooks/internal/useAcknowledgedActivity.js
Comment (corinagum) [reply packages/component/src/hooks/internal/useAcknowledgedActivity.js:20]: Don't see 5. added to the code yet, but looks good.
```suggestion
//      5. Therefore, when the developer sets "pause after 3 activities", if activities are coming in at a slow pace (not batched in a single render)
//         Web Chat will keep scrolling and not become snapped/paused
```

**Comment by Sandy2725** [comment]
Hi @compulim ,

Hope you are doing well.

I have two queries.

1.If we pass the scroll after 1 activity its working fine. But if we have a large content in the 1st message only again the scrollbar is going down. is there any way to focus on at staring of that first message.
2. And if response contains more than (150 or 200 words) or 10 lines. can we add readmore link and then when we click on readmore user can see rest. like after 100 words or 5 lines. for buttons we have readmore but inside the message response do we have that.

Thanks.

**Comment by Sandy2725** [comment]
HI @compulim ,

Can we implement this custom scroll behavior for specific responses. Is there any possibility to implement for only few responses.

styleOptions = {
  autoScrollPauseOnActivity: 2,
  autoScrollPauseOnActivityOffset: 100
};

Thanks .

**Comment by Elephantei** [comment]
After updating the framwork-sdk to `4.12.1` and setting the following `styleOptions` this still doesn't work: 

version: 
`<script src="https://cdn.botframework.com/botframework-webchat/4.12.1/webchat-es5.js" integrity="sha384-YT//LqZJgkEXrmVxm6RZJKs4dSZJQo5kEX0tE9dP1XaOaVsC7NEvwnlP2gW2KWcl" crossorigin="anonymous"></script>`

Styling options added for the chat rendering options:

```
window.WebChat.renderWebChat(
  {
    directLine: window.WebChat.createDirectLine({
      token: process.env.TOKEN,
    }),
    onTelemetry: handleTelemetry,
    store: store,
    styleOptions: {
      hideUploadButton: true,
      autoScrollPauseOnActivity: 2,
      autoScrollPauseOnActivityOffset: 100
    },
  },
  document.getElementById("webchat")
); 
```

Can someone shed some light on it?

**Comment by Sandy2725** [comment]
Hi @Elephantei ,

Try with below style options.

autoScrollSnapOnActivity:2,
autoScrollSnapOnPage:true,
autoScrollSnapOnActivityOffset:100,

**Comment by Jakob-vdH** [comment]
Hi @Sandy2725 

thank you for your help, this worked for us really well for a bit. 
Just to recap, with the following settings, the webchat was always scrolling to the top edge of  a new message / adaptive card, if some messages were in the chat, that were longer that the view, which was awesome behavior.

> autoScrollSnapOnActivity:2,
> autoScrollSnapOnPage:true,
> autoScrollSnapOnActivityOffset:100,

But now we updated to the last webchat version (4.15.2) and this just stopped working, not scrolling anymore at all as soon as the "page" is full with messages. I guess this is intended behavior of `autoScrollSnapOnPage`.
Is  there an option to always scroll to the top of a message if it is longer than the current view?
We tried recreating the initial behavior using the autoScrollSnap options, but sadly did not succeed.

**Pull request merged**
